File: 백준/Gold/2467. 용액/용액.java
import java.util.*;
import java.io.*;

public class Main {
	
	public static void main(String[] args) throws Exception {
		// System.setIn(new FileInputStream("./src/res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = null;
		
		int n = Integer.parseInt(br.readLine());
		long[] arr = new long[n];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i++) {
			arr[i] = Long.parseLong(st.nextToken());
		}
		
		long min = Long.MAX_VALUE;
		
		int ml = 0, mr = 0;
		for (int i = 0; i < n-1; i++) {
			int left = i+1;
			int right = n-1;
			while (left <= right) {
				int mid = (left + right) / 2;
				long sum = Math.abs(arr[i] + arr[mid]);
				
				if (min > sum) {
					min = sum;
					ml = i; mr = mid;
				}
				if (arr[mid] + arr[i] >= 0) {
					right = mid-1;
				} else {
					left = mid+1;
				}
			}
		}
		System.out.println(arr[ml] + " " + arr[mr]);
	}
}

File: 백준/Gold/2467. 용액/용액.js
let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().split("\n");
const n = Number(input[0]);
const arr = input[1].split(" ").map(Number);

let min = Infinity
let resultLeft;
let resultRight;

for (let i = 0; i < n - 1; i++) {
  binarySearch(arr, i, arr[i]);
}

resultLeft < resultRight
  ? console.log(resultLeft, resultRight)
  : console.log(resultRight, resultLeft);

function binarySearch(arr, index, currentNum) {
  let left = index + 1;
  let right = arr.length - 1;
  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    const sum = Math.abs(arr[mid] + currentNum);
    if (min > sum) {
      min = sum;
      resultLeft = arr[mid];
      resultRight = currentNum;
    }

    if (arr[mid] + currentNum >= 0) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }
}
